apps/pano/app/routes/commentEdit.tsx

```diff
@@ -1,17 +1,49 @@
 import type { ActionFunction } from "@remix-run/node";
-import invariant from "tiny-invariant";
+import { json } from "@remix-run/node";
+import { z } from "zod";
 import { editComment } from "~/models/comment.server";
 
+const CommentSchema = z.object({
+  commentID: z.string(),
+  commentContent: z.string().min(1),
+});
+
+type ActionData = {
+  error: {
+    message: string;
+    id: string | null;
+  };
+};
+
+const errorMessage = (message: string, id: string) => {
+  return json<ActionData>(
+    {
+      error: {
+        message,
+        id,
+      },
+    },
+    {
+      status: 400,
+    }
+  );
+};
+type CommentFields = z.infer<typeof CommentSchema>;
+
 export const action: ActionFunction = async ({ request }) => {
   const formData = await request.formData();
-  const jsonData = formData.get("json") as string | null;
+  const jsonData = formData.get("json") as string;
 
-  invariant(jsonData, "Missing json data");
+  const parsedFields = JSON.parse(jsonData) as CommentFields;
+  const result = CommentSchema.safeParse(parsedFields);
 
-  const { commentID, commentContent } = JSON.parse(jsonData);
+  if (!result.success) {
+    return errorMessage("Yorum boş gönderilemez.", parsedFields.commentID);
+  }
 
   try {
-    await editComment(commentID, commentContent);
+    const fields = CommentSchema.parse(parsedFields);
```

Comment:
we don't need to parse this schema again with the parsed fields because we are doing the same operation on line 38. after that validation we know that those fields are valid and ready to go.